executor/radix_hash_join.go

```diff
@@ -0,0 +1,266 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"math"
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/mvmap"
+	log "github.com/sirupsen/logrus"
+	"github.com/spaolacci/murmur3"
+)
+
+var (
+	_ Executor = &RadixHashJoinExec{}
+)
+
+// RadixHashJoinExec implements the hash join algorithm.
+type RadixHashJoinExec struct {
+	*HashJoinExec
+
+	// radixBits indicates the bits using for radix partitioning. Inner relation
+	// will be split to 2^radixBitsNumber sub-relations before building the hash
+	// tables. If the complete inner relation can be hold in L2Cache in which
+	// case radixBits will be 1, we can skip the partition phase.
+	// Note: We actually check whether `size of sub inner relation < 3/4 * L2
+	// cache size` to make sure one inner sub-relation, hash table, one outer
+	// sub-relation and join result of the sub-relations can be totally loaded
+	// in L2 cache size. `3/4` is a magic number, we may adjust it after
+	// benchmark.
+	radixBits  uint32
+	innerParts []partition
+	// innerRowPrts indicates the position in corresponding partition of every
+	// row in innerResult.
+	innerRowPrts [][]partRowPtr
+	// hashTables stores the hash tables built from the inner relation, if there
+	// is no partition phase, a global hash table will be stored in
+	// hashTables[0].
+	hashTables      []*mvmap.MVMap
+	numNonEmptyPart int
```

Comment:
how about reorganizing these fields according to the initialization time? for example:
```go
radixBits
innerRowPtrs

numNonEmptypart
innerParts
hashTables
```

Reply:
I reordered the fields, but not follow the order of initialization time.